faker/providers/barcode/__init__.py

```diff
@@ -23,8 +58,157 @@ def ean(self, length=13):
+          is not set or is invalid, a random 6 digit combination will be used.
+        - If `upc_ae_mode` is disabled, the values of `base` and `number_system_digit` are ignored.
+
+        :param upc_ae_mode: Set to True explicitly to enable
+        :param base: A 6-digit string
+        :param number_system_digit: 0, 1, '0', or '1'
+        :return: 12-digit UPC-A barcode
+        """
+        if upc_ae_mode is True:
+            return self._upc_ae(base=base, number_system_digit=number_system_digit)
+        else:
+            ean13 = self.ean13(force_leading_zero=True)
+            return ean13[1:]
+
+    def upc_e(self, *, base=None, number_system_digit=None, safe_mode=True):
+        """
+        Creates an 8-digit UPC-E barcode
+
+        Notes on UPC-E barcode and this method:
+        - UPC-E barcodes can be expressed in 6, 7, or 8-digit formats, but this method uses the
+          8 digit format, since it is trivial to convert to the other two formats.
+        - The first digit of the barcode denotes the number system used, either 0 or 1.
+        - The last digit is the check digit (more on that below).
+        - The remaining 6 digits are a bit more involved, but they are referred to as the `base`
+          argument elsewhere in this provider (for the lack of a concise and better term).
+        - Each UPC-E `base` have 2 UPC-A equivalents depending on the number system used.
+        - If the value of `number_system_digit` is not 0 or 1, either values will be chosen at random.
+        - A 6-digit string `base` may also be explicitly set. If invalid, it will be ignored, and a
+          new value for `base` will be generated.
+        - Internally, this method first generates a UPC-A barcode that can be converted to UPC-E, and
+          then actually performs a conversion for the result. This is because both the number
+          system and check digits of the UPC-E barcode are inherited from its UPC-A equivalent.
+        - Then there is `safe_mode`. Enabling this guarantees that every UPC-E barcode generated can
+          be converted to UPC-A, and that UPC-A value can be converted back again to UPC-E and still
+          be equal to the original value. In other words, the statement a2e(e2a(e)) == e holds True.
+        - As to why there is `safe_mode`, it is because there are some UPC-E values that share the
+          same UPC-A equivalent. For example, any UPC-E barcode of the form abc0000d, abc0003d, and
+          abc0004d share the same UPC-A value abc00000000d, but that UPC-A value will only convert
+          to abc0000d because of (a) how UPC-E is just a zero-suppressed version of UPC-A and
+          (b) the rules around the conversion.
 
-    def ean13(self):
-        return self.ean(13)
+        :param base: A 6-digit string
+        :param number_system_digit: First digit of the barcode
+        :param safe_mode: True or False
```

Comment:
Could you add a description about `safe_mode` here too? I'm thinking people scanning through the code would probably expect to find it here.

Reply:
There is actually an explanation in the docstring, but I honestly do not know how to describe it concisely. Maybe you can give me suggestions.

Basically, in accordance to a set of rules, there are UPC-A barcodes that may be zero-suppressed (kind of like how some zeroes may be omitted in IPv6), and those have a UPC-E equivalent. The problem is that if you generate certain UPC-E barcodes arbitrarily out of thin air and convert those to UPC-A and back again to UPC-E, there is a possibility the new UPC-E barcode will not be the same as the original.

You can try that quirk [here](https://www.morovia.com/education/utility/upc-ean.asp). The converter accepts a 6 digit UPC-E base, so try converting `100003` to UPC-A. Copy paste the result, and convert it back to UPC-E, and you will get `100000`.

Safe mode basically guarantees that every UPC-E barcode generated, even if such a base was provided explicitly, can be converted to UPC-A and back to UPC-E ad infinitum.

Reply:
I think something like this would work:

```
    :param safe_mode: True or False. Guarantees that every UPC-E barcode generated can be converted back-and forth between UPC-A and UPC-E.
```

Reply:
I had something like that in mind, but I felt it was still long. As long as it has your blessing, then I am all for it as well. :+1: